src/collections.rs

```diff
@@ -386,11 +527,17 @@ pub unsafe extern "C" fn z_string_wrap(
 #[no_mangle]
 #[allow(clippy::missing_safety_doc)]
 pub unsafe extern "C" fn z_string_from_substring(
-    this: *mut MaybeUninit<z_owned_string_t>,
+    this: &mut MaybeUninit<z_owned_string_t>,
     str: *const libc::c_char,
     len: usize,
 ) -> z_error_t {
-    z_slice_wrap(this as *mut _, str as _, len)
+    match CStringOwned::new_owned(str, len) {
+        Ok(slice) => {
+            this.as_rust_type_mut_uninit().write(slice);
+            errors::Z_OK
+        }
+        Err(e) => e,
```

Comment:
same as for slice, we still need to initialize string to some defined value even in case of error.